File: content/examples/other/leetcode/problems/course-schedule/example-2/solution.js
function canFinishDFS(numCourses, prerequisites) {
  const adj = Array.from({ length: numCourses }, () => []);
  for (const [course, prereq] of prerequisites) adj[prereq].push(course);

  // 0 = unvisited, 1 = visiting, 2 = visited
  const state = new Array(numCourses).fill(0);

  function dfs(node) {
    if (state[node] === 1) return false; // cycle
    if (state[node] === 2) return true;
    state[node] = 1;
    for (const nxt of adj[node]) {
      if (!dfs(nxt)) return false;
    }
    state[node] = 2;
    return true;
  }

  for (let i = 0; i < numCourses; i += 1) {
    if (!dfs(i)) return false;
  }
  return true;
}

if (require.main === module) {
  console.log(canFinishDFS(2, [[1, 0]]));
  console.log(canFinishDFS(2, [[1, 0], [0, 1]]));
}

module.exports = { canFinishDFS };

File: content/examples/other/leetcode/problems/course-schedule/example-2/solution.py
def can_finish_dfs(num_courses, prerequisites):
    adj = [[] for _ in range(num_courses)]
    for course, prereq in prerequisites:
        adj[prereq].append(course)

    # 0=unvisited, 1=visiting, 2=visited
    state = [0] * num_courses

    def dfs(node):
        if state[node] == 1:
            return False
        if state[node] == 2:
            return True
        state[node] = 1
        for nxt in adj[node]:
            if not dfs(nxt):
                return False
        state[node] = 2
        return True

    for i in range(num_courses):
        if not dfs(i):
            return False
    return True


if __name__ == "__main__":
    print(can_finish_dfs(2, [(1, 0)]))
    print(can_finish_dfs(2, [(1, 0), (0, 1)]))
